### ISSUE TITLE
Wrong results for Cypher query after renaming variable

### ISSUE THREAD

**Issue by matiasmorant** [neo4j/neo4j#12669]
I discovered a Cypher query that produces wrong results. It returns a wrong value for the property of a node. This seems to be related renaming variables

**Neo4j version:** 4.2.3
**Operating system:** Windows 10
**API/Driver:** Cypher

### Steps to reproduce

1- create a new graph:
```
CREATE (:Person {name:'A'})-[:KNOWS]->(:Person {name:'B'})
```
2- run this query:
```
MATCH (x {name:'A'})
CALL {
    WITH x RETURN x AS y
    UNION
    WITH x MATCH (x)-[]-(y) RETURN y
}
WITH y AS z
RETURN z.name, id(z)
```

### Expected behavior
```
╒════════╤═══════╕
│"z.name"│"id(z)"│
╞════════╪═══════╡
│"A"     │0      │
├────────┼───────┤
│"B"     │1      │
└────────┴───────┘
```
### Actual behavior 
```
╒════════╤═══════╕
│"z.name"│"id(z)"│
╞════════╪═══════╡
│"A"     │0      │
├────────┼───────┤
│"A"     │1      │
└────────┴───────┘
```
note it gets the wrong name for node B

### Hint

If we remove the last renaming (`AS z`), actual behavior matches expected behavior:

```
MATCH (x {name:'A'})
CALL {
    WITH x RETURN x AS y
    UNION
    WITH x MATCH (x)-[]-(y) RETURN y
}
WITH y
RETURN y.name, id(y)
```

**Comment by mnd999** [comment]
Thanks for the report. That certainly looks like something is wrong at first glance. We'll investigate further and let you know what we find.

**Comment by SaschaPeukert** [comment]
Hej @matiasmorant, 

Once again thank you for reporting this issue. We have fixed the problem in the already released version 4.2.4 and it will also be part of the upcoming release of 4.1.9

I'll close this now but don't hesitate to reopen or open a new issue if there are any other issues.

Regards,
Sascha
